align_system/algorithms/outlines_hybrid_regression_adm.py

```diff
@@ -0,0 +1,237 @@
+                                distribution_matching='sample',
+                                kde_norm='globalnorm',
+                                **kwargs):
+
+        scenario = []
+        for action in available_actions:
+            for character in scenario_state.characters:
+                chosen_character = None
+                if action.character_id is None:
+                    continue
+                if character.id == action.character_id:
+                    chosen_character = character
+                    break
+            scenario_description = (
+                scenario_description_hybrid_regression(
+                    scenario_state,
+                    chosen_character
+                )
+            )
+            scenario.append(scenario_description)
+
+        # Important that the choices stay in the same order as the
+        # available actions as we'll use the selected index later to
+        # map to the corresponding action
+        choices = self.format_choices(
+            [a.unstructured for a in available_actions],
+            available_actions,
+            scenario_state
+        )
+
+        data_dict = {
+            'scenario': scenario,
+            'answer': choices
+        }
+
+        target_models = kwargs.get('models', {})
+        if not target_models:
+            raise KeyError("'models' field is empty")
+        for k in ["target_checkpoint", "model_name"]:
+            if k not in target_models:
+                raise KeyError(f"{k} does not exist")
+
+        target_model_checkpoints = target_models["target_checkpoint"].keys()
+        model_name = target_models["model_name"]
+
+        target_kdmas = alignment_target.kdma_values
+
+        for i, target_kdma in enumerate(target_kdmas):
+            target_kdma = dict(target_kdma)
+            # Only works for scalar alignment targets
+            target_kdma_name = target_kdma["kdma"]
+            target_kdma_value = target_kdma["value"]
+
+            if target_kdma_name not in target_model_checkpoints:
+                raise ValueError(f"Model checkpoint does not exist for {target_kdma_name}")
+            log.info(
+                f"Chosen KDMA: {target_kdma_name}, "
+                f"Target value: {target_kdma_value}"
+            )
+            predictions_per_kdma = get_kdma_predictions(
+                model_name=model_name,
+                data_dict=data_dict,
+                model_path=target_models["target_checkpoint"][str(target_kdma_name)]
+            )
+            predicted_kdma_values = {}
+            log.info("ACTION CHOICES AND REGRESSION MODEL PREDICTIONS:")
+            for pred, choice in zip(predictions_per_kdma, choices):
+                # Only works for scalar alignment targets
+                dist = (target_kdma_value - pred)**2
+                predicted_kdma_values.update({
+                    choice: {
+                        target_kdma_name: pred
+                    }
+                })
```

Comment:
I just tested this and it looks like it would be overwritten, I **think** you want to do something like

```
if choice not in predicted_kdma_values:
    predicted_kdma_values[choice] = {}
predicted_kdma_values[choice][target_kdma_name] = pred
```